File: src/main/java/org/distril/beengine/network/data/LoginData.java
package org.distril.beengine.network.data;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.nimbusds.jose.JWSObject;
import com.nimbusds.jose.shaded.json.JSONArray;
import com.nukkitx.protocol.bedrock.data.skin.SerializedSkin;
import com.nukkitx.protocol.bedrock.util.EncryptionUtils;
import io.netty.util.AsciiString;
import lombok.Getter;
import org.distril.beengine.util.SkinUtil;

import java.util.UUID;

@Getter
public class LoginData {

	private static final Gson GSON = new Gson();

	private final String xuid;
	private final String identityPublicKey;
	private final UUID uuid;
	private final String username;
	private final String languageCode;
	private final Device device;
	private final SerializedSkin skin;
	private final boolean authenticated;


	private LoginData(String xuid, String identityPublicKey, UUID uuid, String username, String languageCode,
	                  Device device, SerializedSkin skin, boolean authenticated) {
		this.xuid = xuid;
		this.identityPublicKey = identityPublicKey;
		this.uuid = uuid;
		this.username = username;
		this.languageCode = languageCode;
		this.device = device;
		this.skin = skin;
		this.authenticated = authenticated;
	}

	public static LoginData extract(AsciiString chainData, AsciiString skinData) {
		try {
			JsonObject chainJSON = GSON.fromJson(chainData.toString(), JsonObject.class);

			JSONArray chains = new JSONArray();
			chainJSON.getAsJsonArray("chain").forEach(chain -> chains.add(chain.getAsString()));

			var authenticated = EncryptionUtils.verifyChain(chains);

			// Retrieve xuid, uuid, and username
			String chainPayload = JWSObject.parse(((String) chains.get(chains.size() - 1))).getPayload().toString();
			JsonObject jsonPayload = GSON.fromJson(chainPayload, JsonObject.class);
			JsonObject extraData = jsonPayload.getAsJsonObject("extraData");

			var xuid = extraData.get("XUID").getAsString();
			var uuid = UUID.fromString(extraData.get("identity").getAsString());
			var username = extraData.get("displayName").getAsString();
			var identityPublicKey = jsonPayload.get("identityPublicKey").getAsString();

			// Extract data from skin string
			JWSObject skinJWS = JWSObject.parse(skinData.toString());
			JsonObject skinJSON = GSON.fromJson(skinJWS.getPayload().toString(), JsonObject.class);

			// Retrieve device and language code
			var device = Device.fromOS(skinJSON.get("DeviceOS").getAsInt());
			var languageCode = skinJSON.get("LanguageCode").getAsString();

			// Retrieve skin
			var skin = SkinUtil.fromToken(skinJSON);
			return new LoginData(xuid, identityPublicKey, uuid, username, languageCode, device, skin, authenticated);
		} catch (Exception exception) {
			return null;
		}
	}
}

File: src/main/kotlin/org/distril/beengine/network/data/LoginData.kt
package org.distril.beengine.network.data

import com.google.gson.JsonObject
import com.nimbusds.jose.JWSObject
import com.nimbusds.jose.shaded.json.JSONArray
import com.nukkitx.protocol.bedrock.data.skin.SerializedSkin
import com.nukkitx.protocol.bedrock.util.EncryptionUtils
import io.netty.util.AsciiString
import org.distril.beengine.util.SkinUtil
import org.distril.beengine.util.Utils.gson
import java.util.*

class LoginData private constructor(
    val xuid: String,
    val identityPublicKey: String,
    val uuid: UUID,
    val username: String,
    val languageCode: String,
    val device: Device,
    val skin: SerializedSkin,
    val authenticated: Boolean
) {

    companion object {

        fun extract(chainData: AsciiString, skinData: AsciiString): LoginData? {
            return try {
                val chainJSON = gson.fromJson(chainData.toString(), JsonObject::class.java)
                val chains = JSONArray()
                chainJSON.getAsJsonArray("chain").forEach { chains.add(it.asString) }

                val authenticated = EncryptionUtils.verifyChain(chains)

                // Retrieve xuid, uuid, and username
                val chainPayload = JWSObject.parse(chains[chains.size - 1] as String).payload.toString()
                val jsonPayload = gson.fromJson(chainPayload, JsonObject::class.java)
                val extraData = jsonPayload.getAsJsonObject("extraData")
                val xuid = extraData["XUID"].asString
                val uuid = UUID.fromString(extraData["identity"].asString)
                val username = extraData["displayName"].asString
                val identityPublicKey = jsonPayload["identityPublicKey"].asString

                // Extract data from skin string
                val skinJWS = JWSObject.parse(skinData.toString())
                val skinJSON = gson.fromJson(skinJWS.payload.toString(), JsonObject::class.java)

                // Retrieve device and language code
                val device = Device.fromOSId(skinJSON["DeviceOS"].asInt)
                val languageCode = skinJSON["LanguageCode"].asString

                // Retrieve skin
                val skin = SkinUtil.fromToken(skinJSON)
                LoginData(xuid, identityPublicKey, uuid, username, languageCode, device, skin, authenticated)
            } catch (exception: Exception) {
                null
            }
        }
    }
}
